### PR TITLE
[IDE][Index] Property wrapper rename support

### PR DESCRIPTION
**Pull request by nathawes**
This PR fixes up rename support for property wrappers. There were three main issues this addresses:

1. Renaming an `@propertyWrapper` struct was failing due to rename's `NameMatcher` using the range of a `Decl` itself, not including its attributes, to decide if it contained a location that needed to be resolved.
2. Renaming a constructor of a `@propertyWrapper` struct would not pick up occurrences of that constructor's argument labels used in custom attributes, e.g. renaming `MyWrapper`'s `init(initialValue:minValue:maxValue)` constructor would miss the argument labels in custom attributes like `@MyWrapper(initialValue: 50, minValue: 0, maxValue: 100)`. This is because the indexer didn't report an occurrence of the constructor call in custom attributes. Rename's `NameMatcher` also needed to be updated to handle resolving these newly-added occurrences, since the call expression in the AST doesn't have an explicit base (`MyWrapper` is not part of the call expression, but is where the index reports the call taking place).
3. Renaming a wrapped property (e.g. `foo`) wasn't updating the corresponding synthesized `$`-prefixed backing property (`$foo`). The synthesized backing property declarations were implicit, so the index wasn't recording them, and even if it did, it's a different property so wouldn't be picked up automatically.
\
To resolve this, the index now records an occurrence of the wrapped property within the synthesized backing property occurrences (i.e. on the `foo` within each `$foo`) so that rename on the wrapped property updates the synthesized backing property occurrences too.  To support rename being invoked via an occurrence of a synthesized property, the `CusorInfo` request that determines which symbol is renamed for a given location now treats occurrences of synthesized properties as if they were the corresponding wrapped properties. The `FindRelatedIdents` request was also similarly updated so that rename-like features that are based on it (like Xcode's edit-all-in-scope) can get both the underlying and synthesized property ranges, regardless of which is was invoked on.

This PR also includes some small and/or incidental IDE issue fixes around custom attributes:
- syntax coloring wasn't working for the custom attribute type occurrences (only semantic highlighting)
- interleaving of availability attributes, custom attributes and built-in attributes would sometimes result in incorrect syntax highlighting.

Resolves rdar://problem/49036613
Resolves rdar://problem/50073641
Resolves rdar://problem/51695783
Resolves rdar://problem/52053678

### CODE REVIEW

**Comment by nathawes** [comment]
@swift-ci please test

**Comment by nathawes** [comment]
@swift-ci please test

File: lib/IDE/SwiftSourceDocInfo.cpp [hunk 1]
```diff
@@ -354,6 +354,39 @@ static std::vector<CharSourceRange> getEnumParamListInfo(SourceManager &SM,
   return LabelRanges;
 }
 
+bool NameMatcher::handleCustomAttrs(Decl *D) {
+  // CustomAttrs of non-param VarDecls are handled by their containing
+  // PatternBindingDecl
```
Comment (benlangmuir) [inline lib/IDE/SwiftSourceDocInfo.cpp]: Should add this comment to `SemaAnnotator::handleCustomAttributes` as well.

File: lib/Index/Index.cpp [hunk 2]
```diff
@@ -450,6 +477,16 @@ class IndexSwiftASTWalker : public SourceEntityWalker {
     return nullptr;
   }
 
+  Decl *getContainingDecl() const {
+    auto Containers = (SymbolRoleSet)SymbolRole::Definition | (SymbolRoleSet)SymbolRole::Declaration;
+    for (const auto &Entity: EntitiesStack) {
```
Comment (benlangmuir) [inline lib/Index/Index.cpp]: When does a declaration (not a definition) contain something?  Is this for default argument expressions or something?

File: test/Index/property_wrappers.swift [hunk 3]
```diff
@@ -18,25 +18,38 @@ var globalInt: Int { return 17 }
 public struct HasWrappers {
   @Wrapper
   public var x: Int = globalInt
+  // CHECK: [[@LINE-2]]:4 | struct/Swift | Wrapper | s:14swift_ide_test7WrapperV | Ref | rel: 0
+  // CHECK-NOT: [[@LINE-2]]:23 | variable/Swift | globalInt | s:14swift_ide_test9globalIntSivp | Ref,Read | rel: 0
```
Comment (benlangmuir) [inline test/Index/property_wrappers.swift]: I suggest dropping everything after the name `globalInt` to make this CHECK-NOT more robust against changes.

File: test/refactoring/SyntacticRename/FindRangeOutputs/property-wrapper-init/body.swift.expected [hunk 4]
```diff
@@ -0,0 +1,37 @@
+public struct Outer {
+    @propertyWrapper
+    public struct InnerWrapper<T> {
+        public var value: T
+        public /*init:def*/init(initialValue: T) {
```
Comment (benlangmuir) [inline test/refactoring/SyntacticRename/FindRangeOutputs/property-wrapper-init/body.swift.expected]: Do you have a test somewhere for `init(initialValue: T, other: T)` with `@Wrapper(other: 1) var foo: Int = 2`?

Comment (nathawes) [reply test/refactoring/SyntacticRename/FindRangeOutputs/property-wrapper-init/body.swift.expected]: Oh thanks – I didn't realize that was in yet. I think the initializer rename should still be fine, but given the passed arguments could be closure bodies and can now be split between the custom attribute and the var initializer, NameMatcher and SyntaxModel may have issues with things being out of source order...

File: lib/Index/Index.cpp (hunk 2 above)
Comment (nathawes) [reply lib/Index/Index.cpp]: Nope, this just doesn't make sense 🙂
I don't even think we use the Declaration role in swift...

File: test/refactoring/SyntacticRename/FindRangeOutputs/property-wrapper-init/body.swift.expected (hunk 4 above)
Comment (nathawes) [reply test/refactoring/SyntacticRename/FindRangeOutputs/property-wrapper-init/body.swift.expected]: Added tests and it looks like they're actually fine. They're running over the non-typechecked AST, so they're still two distinct pieces.

**Comment by nathawes** [comment]
@swift-ci please test

**Review by benlangmuir — APPROVED** [review]

**Comment by nathawes** [comment]
@swift-ci please test

**Comment by nathawes** [comment]
@swift-ci please test

**Comment by nathawes** [comment]
@swift-ci please test

**Comment by nathawes** [comment]
@swift-ci please test OS X Platform

**Review by benlangmuir — APPROVED** [review]

**Pull request merged**